ros_buildfarm/config/doc_build_file.py

```diff
@@ -167,14 +167,21 @@ def __init__(self, name, data):  # noqa: D107
         # user host and docroot have default of uploading to the repo machine
         # next to the debs
         self.upload_user = data.get('upload_user', 'jenkins-agent')
-        self.upload_host = data.get('upload_host', 'repo')
+        self.upload_host = data.get('upload_host', None)
         self.upload_root = data.get('upload_root', '/var/repos/docs')
+
+        self.upload_repository_url = None
         if self.documentation_type == DOC_TYPE_DOCKER:
-            assert 'upload_repository_url' in data
-            self.upload_repository_url = data['upload_repository_url']
+            self.upload_repository_url = data.get('upload_repository_url', None)
             self.upload_repository_branch = data.get(
                 'upload_repository_branch', 'gh-pages'
             )
+
+            self.doc_repository_branch = data.get('doc_repository_branch', None)
```

Comment:
This is a new user-visible configuration parameter.  It allows the user to specify what branch to checkout on the source documentation repository.  I know I'll need to document this, but I wanted to make sure that adding something like this is OK before going further.

Reply:
My primary concern is that doc_repositories is a list of repositories while the proposed branch option is singular.

The naive solution would be to follow the pattern of repository urls and keys and make `doc_repository_branches` an array where `url[0]` corresponds to `branch[0]` but I don't care much for the obfuscated connection there and would much rather support a compound data structure such as `doc_repositories:{url:,branch:} or even adopting the full `doc_repositories:{type:,url:,version:}` structure for repositories that is used in rosdistro distribution.yaml. I'd value @cottsay's opinion on the matter as well since the fact that this peeves me may mean it weighs more than it should in my review.

Reply:
Yes, that actually annoyed me too.  And your concern about the obfuscated connection there is exactly why I didn't go that route.

I'd be for making a compound structure like you mention above, but I couldn't quite see how to do it in a backwards-compatible way.  But that may just be my unfamiliarity with YAML.  If someone could suggest a structure that would be backwards-compatible, I'd be happy to implement it that way.

Reply:
Looking at this closer, I could do the following.  If the `doc_repositories` section looks like:

```
doc_repositories:
- https://github.com/clalancette/ros2_documentation.git
```

Then the Python type will be list.  If it looks more like:

```
doc_repositories:
  ros2_doc:
    url: https://github.com/clalancette/ros2_documentation.git
    branch: clalancette/rolling-add-multiversion-target
```

Then the Python type will be dict.  So I can do something like the following in the Python code:

```
repos = {}
if isinstance(data['doc_repositories'], list):
    for url in data['doc_repositories']:
        repos[url] = ''
elif isinstance(data['doc_repositories'], dict):
    for repo in data['doc_repositories']:
        assert 'url' in data['doc_repositories'][repo]
        url = data['doc_repositories'][repo]['url']
        if 'branch' in data['doc_repositories'][repo]:
            repos[url] = data['doc_repositories'][repo]['branch']
        else:
            repos[url] = ''
```

Which would always get me a dictionary from 'url' -> 'branch' (and where the branch name may be empty, indicating use the default branch).  What do you think about that?

Reply:
Just to be clear; I don't think this would break the API, so much as *add* to the API.

That is, if we go this way, then what worked in the YAML before (a list of doc_repositories) would still work.  It's just that we would additionally support having a dictionary of values in a particular way in the YAML.

I'm going to go ahead and prototype this and see what it ends up looking like.

Reply:
All right, I've coded this up and tested it on the test farm, and things seem to be good.  I've also documented it now.  So I'm going to mark this conversation resolved, and change this to a "real" PR.